Support scan filter for ORC decimal reader

rui-mo opened the pull request:
'leafCallToSubfieldFilter' converts decimal filter as Subfield filter, but 
'SelectiveDecimalColumnReader::read' rejects scan spec filter. This PR supports 
scan filter for ORC decimal reader.

Yuhta commented on velox/expression/ExprToSubfieldFilter.cpp:
```diff
@@ -512,6 +512,13 @@ std::unique_ptr<common::Filter> leafCallToSubfieldFilter(
       }
       return isNull();
     }
+  } else if (call.name() == "isnotnull") {
```
Isn't `IS NOT NULL` parsed into `not(is_null(...))`?

rui-mo replied:
In Spark there is an expression 'IsNotNull' and it is frequently used in filter pushdown.This issue adds the background for this change: https://github.com/facebookincubator/velox/issues/11093. Thanks.

rui-mo commented:
> You also need to add some tests to E2EFilterTest for decimal types

@Yuhta I find the E2EFilterTest relies on DWRF writer, but for now bigint-decimal will be written as int64_t, and hugeint-decimal is not supported. Do we need to support them first?

https://github.com/facebookincubator/velox/blob/3a1a60af69ad9d2f9060ce0c22aa212224fd6053/velox/dwio/dwrf/writer/ColumnWriter.cpp#L2039-L2041

https://github.com/facebookincubator/velox/blob/3a1a60af69ad9d2f9060ce0c22aa212224fd6053/velox/dwio/dwrf/writer/ColumnWriter.cpp#L2090

Yuhta commented:
@rui-mo Yes it would be ideal to support writing decimals first.  Otherwise the tests are very limited and we have no confidence the new code would work correctly.

zhouyuan commented on velox/exec/tests/TableScanTest.cpp:
```diff
@@ -1838,6 +1841,124 @@ TEST_F(TableScanTest, validFileNoData) {
+
+  auto filePath = facebook::velox::test::getDataFilePath(
+      "velox/exec/tests", "data/short_decimal.orc");
+  auto split = HiveConnectorSplitBuilder(filePath)
```
should replace with "exec::test::HiveConnectorSplitBuilder" ?

rui-mo commented on velox/dwio/dwrf/reader/SelectiveDecimalColumnReader.cpp:
```diff
@@ -96,46 +97,175 @@ void SelectiveDecimalColumnReader<DataT>::readHelper(RowSet rows) {
+  // Fill decimals before applying filter.
+  fillDecimals();
+
+  const auto rawNulls = nullsInReadRange_
```
Parquet decimal uses `SelectiveIntegerColumnReader::processFilter` so it might don't need to reuse this logic. How do you think? Thanks.

aditi-pandit commented on velox/dwio/dwrf/reader/SelectiveDecimalColumnReader.cpp:
```diff
@@ -96,46 +97,194 @@ void SelectiveDecimalColumnReader<DataT>::readHelper(RowSet rows) {
+    const common::Filter* filter,
+    const RowSet& rows,
+    const uint64_t* rawNulls) {
+  returnReaderNulls_ = false;
```
Add VELOX_CHECK(filter) to assert the function is always called with it.

rui-mo commented:
> @rui-mo Yes it would be ideal to support writing decimals first. Otherwise the tests are very limited and we have no confidence the new code would work correctly.

The E2E filter tests depend on decimal write support in https://github.com/facebookincubator/velox/pull/11431.

majetideepak commented:
@rui-mo do we still need to check in the orc files if we now have the writer?

rui-mo commented:
Hi @majetideepak, I removed the orc file tests and added E2E filter tests. I find current footer writing is incompatible with orc file format, seeing details in https://github.com/facebookincubator/velox/issues/12418. To enable the orc decimal E2E tests, some status adjustment only for testing are added in `ReaderBase::convertType` and `ProtoUtils::writeType` as a workaround for the footer difference. Do you have any suggestion? Thanks! cc: @Yuhta

majetideepak commented:
@rui-mo I feel if we use the proto::orc::Footer for writes, we can then do the right thing and generate valid ORC files?

rui-mo commented:
@Yuhta @majetideepak I integrated this PR with the ORC writer support in https://github.com/facebookincubator/velox/pull/12487 and the E2E filter tests work, seeing https://github.com/rui-mo/velox/commit/f04fbfc94dbee6395d6fa26fb83b0b16718b17cd. Would you rather we merge this PR first, or should I wait for the ORC writer to be landed? This PR contains a few workarounds on the footer for testing purposes only, as the ORC writer is not supported.

Yuhta commented:
@rui-mo Let's merge the writer change first, the E2EFilterTest won't be much useful if not running in optimized mode.

pedroerp commented:
I'll defer this to @Yuhta, but agreed with the sentiment that having a writer and some fuzzed e2e tests would be better. 

Looks like that https://github.com/facebookincubator/velox/pull/12487 was abandoned. Any plans on picking it up again?

rui-mo commented:
> I'll defer this to @Yuhta, but agreed with the sentiment that having a writer and some fuzzed e2e tests would be better.

Thanks, @pedroerp. I agree that having e2e tests would be better. Since ORC writer support seems to be a longer-term effort, would it make sense for me to open an issue to track the e2e test for this change and then add it once ORC writer support is merged? I’d be glad to hear your thoughts.
By the way, I’ve already picked up the ORC writer support on my end and verified the e2e test.

pedroerp commented:
Thanks @rui-mo. @Yuhta, considering #12616 is blocked on us, do you think we could merge this one with only unit tests, and add e2e fuzzer tests as a follow up once the writer support is finalized?

rui-mo commented:
> Thanks @rui-mo. @Yuhta, considering #12616 is blocked on us, do you think we could merge this one with only unit tests, and add e2e fuzzer tests as a follow up once the writer support is finalized?

Hi @Yuhta, I’d appreciate hearing your thoughts on this proposal. Thanks!

Yuhta commented on velox/dwio/dwrf/reader/ReaderBase.cpp:
```diff
@@ -344,13 +344,19 @@ std::shared_ptr<const Type> ReaderBase::convertType(
+            DECIMAL(type.getOrcPtr()->precision(), type.getOrcPtr()->scale());
+      } else {
+        converted = BIGINT();
+        common::testutil::TestValue::adjust(
```
Can you add a comment why this is needed?  Is there a way to avoid this hack?

rui-mo replied:
Hi @Yuhta, I really appreciate your time and feedback on this PR. Let me clarify the context a bit. These hacks are only needed for the E2E filter test, which writes ORC files and read them back. Currently Velox only supports DWRF footer and as I mentioned at https://github.com/facebookincubator/velox/pull/11067#issuecomment-2674372810, the footer of DWRF and ORC are incompatible and https://github.com/facebookincubator/velox/pull/12487 is working on the ORC writer support. I believe there are the below three options to proceed this PR:

(1) The ideal way is to land https://github.com/facebookincubator/velox/pull/12487 first, and the E2E filter test can work without any hack. However, the support of ORC writer seems to be a long-term effort.
(2) If the E2E filter test is required now, these hacks are needed to make the test work.
(3) If you agree that we track the E2E test in an issue and add it once ORC writer is supported, we can remove these hacks.

My preference would be option (3), and I’d like to remove these hacks from the implementation. Kindly let me know what you think, thanks!

Yuhta replied:
If https://github.com/facebookincubator/velox/pull/12616 gets merged, can we do proper writes with ORC?  It will be merged very soon (probably this week).

Yuhta replied:
OK let's go with option (3) then, let's create an issue to track this.

Yuhta replied:
Let's also keep the E2E test code ready and run it once right before merging this PR.

Yuhta commented on velox/dwio/dwrf/utils/ProtoUtils.cpp:
```diff
@@ -57,9 +57,22 @@ void ProtoUtils::writeType(
-  self->set_kind(kind);
+  bool kindSet = false;
+  if (type.kind() == TypeKind::HUGEINT) {
+    // Hugeint is not supported by DWRF, and this branch is only for ORC
```
The decimal writer is expected to be merged soon, can we leverage that instead of relying on this hack?

rui-mo replied:
I added the decimal ORC writer https://github.com/facebookincubator/velox/commit/da6c05f20a73b641b3bb21a10d6ab42d9a2e4c10 several months ago, but the footer is still incompatible, seeing https://github.com/facebookincubator/velox/issues/12418.

Yuhta commented on velox/dwio/dwrf/reader/SelectiveDecimalColumnReader.cpp:
```diff
@@ -104,46 +105,201 @@ void SelectiveDecimalColumnReader<DataT>::readHelper(RowSet rows) {
-  VELOX_CHECK(!scanSpec_->filter());
   VELOX_CHECK(!scanSpec_->valueHook());
   prepareRead<int64_t>(offset, rows, incomingNulls);
+  if (!resultNulls_ || !resultNulls_->unique() ||
```
This should be allocated in `prepareRead()`->`prepareNulls()`, you should not need to do it yourself here.  In some cases we reuse the incoming nulls directly (when `returnReaderNulls_ = true`) so this should be left as `nullptr`.  Be sure you respect that.

rui-mo replied:
This follows https://github.com/facebookincubator/velox/commit/5bd10e9757b26aff4f120e79d445e9d7557359fd, which fixes the case when the column is not projected out, `rawResultNulls_` is not initialized.

Yuhta replied:
It is not implemented in optimal way in `SelectiveTimestampColumnReader` either, but decimal reader has a bigger blast radius (more usage in data warehouse) so we must be more careful here.  If the column is not projected out, we do not really need the result nulls, and the allocations and writes are perf loss.  Can we figure out which rows are null from the reader nulls (`nullsInReadRange_`) instead?

rui-mo replied:
The issue occurs in a query such as `SELECT b FROM tmp WHERE a IS NULL`, where the filter column a is not included in the projection. In this case, `scanSpec_->keepValues()` returns false, so `prepareNulls()` is not invoked.

https://github.com/facebookincubator/velox/blob/dbaf08b53f910f009ca98d634f0c27e8ec87aa7f/velox/dwio/common/SelectiveColumnReaderInternal.h#L89-L92

I updated this PR to allocate nulls only when `DictionaryValues::hasFilter(scanSpec_->filter())` returns true to reduce unnecessary overhead, and added a test to cover this. Do you think this change makes sense? Thanks.

https://github.com/facebookincubator/velox/pull/11067/files#diff-8927e779a767d849f591e2a6678bc74654547535142720fc8908fbca86a53b29R246-R253

https://github.com/facebookincubator/velox/pull/11067/files#diff-c62590f29a2bf22e50e276f2665e0a7666968880de73a6b4db745cbc0cfde6feR6301-R6305

Yuhta commented on velox/dwio/dwrf/reader/SelectiveDecimalColumnReader.cpp:
```diff
@@ -104,46 +105,201 @@ void SelectiveDecimalColumnReader<DataT>::readHelper(RowSet rows) {
+  if (!rawNulls) {
+    return;
+  }
+  returnReaderNulls_ = false;
```
We don't need nulls buffer in case `isNull = false`

rui-mo replied:
If rawNulls is nullptr and the filter is IsNotNull, all rows will be added to the output directly.

https://github.com/facebookincubator/velox/pull/11067/files#diff-8927e779a767d849f591e2a6678bc74654547535142720fc8908fbca86a53b29R221-R228

Yuhta commented on velox/dwio/dwrf/reader/SelectiveDecimalColumnReader.cpp:
```diff
@@ -104,46 +105,201 @@ void SelectiveDecimalColumnReader<DataT>::readHelper(RowSet rows) {
+    case common::FilterKind::kNegatedBigintValuesUsingBitmask:
+    case common::FilterKind::kBigintMultiRange: {
+      if constexpr (std::is_same_v<DataT, int64_t>) {
+        processFilter(filter, rows, rawNulls);
```
Why do we have to differentiate these `FilterKind` if they are going to call the same `processFilter` (no template specializations)?  Just put everything in `default` should be enough.

Yuhta replied:
Yeah `DataT` is a parameter on the class, the `processFilter` is still the same no matter which filter kind it is.

rui-mo commented on velox/dwio/dwrf/reader/SelectiveDecimalColumnReader.cpp:
```diff
@@ -104,46 +105,201 @@ void SelectiveDecimalColumnReader<DataT>::readHelper(RowSet rows) {
+  fillDecimals();
+
+  const auto rawNulls = nullsInReadRange_
+      ? (kDense ? nullsInReadRange_->as<uint64_t>() : rawResultNulls_)
```
Apologies, I’m not entirely sure I understood correctly. Are you suggesting that we should always pass `nullsInReadRange_`? I made that change, but it led to incorrect results. My understanding was that it could be referenced by result if all rows are selected. This behavior follows the implementation shown below.

https://github.com/facebookincubator/velox/blob/17d47a9c508460cbeef9a1c768ba988b78219ff8/velox/dwio/dwrf/reader/SelectiveTimestampColumnReader.cpp#L141-L143.

Yuhta replied:
Yes we should use `nullsInReadRange_` whenever we can, because result nulls are not always populated as I explained in the other comment.  The access pattern of nulls in `processNulls` and `processFilter` needs to change as well in order for this to work.

rui-mo replied:
Added the below comments, thanks.

> // 'nullsInReadRange_' is the nulls for the entire read range, and if the row
  // set is not dense, result nulls should be allocated, which represents the
  // nulls for the selected rows before filtering.

rui-mo commented on velox/dwio/dwrf/reader/SelectiveDecimalColumnReader.cpp:
```diff
@@ -104,46 +105,177 @@ void SelectiveDecimalColumnReader<DataT>::readHelper(RowSet rows) {
   VELOX_CHECK(!scanSpec_->valueHook());
   prepareRead<int64_t>(offset, rows, incomingNulls);
+  if (!scanSpec_->keepValues() &&
+      DictionaryValues::hasFilter(scanSpec_->filter()) &&
```
`DictionaryValues::hasFilter` prevents null buffer allocation for `IsNotNull` filters, but I've made this change in case you prefer it. Thanks.

rui-mo commented:
@kagamiori This PR is ready-to-merge. Could you please help import it? Thanks!

FelixYBW commented:
Thank you, @Yuhta and @kagamiori. It's one of the oap PR in gluten velox backend.

kagamiori commented:
> Thank you, @Yuhta and @kagamiori. It's one of the oap PR in gluten velox backend.

@FelixYBW Sure thing. Curious what does "oap" represent?

rui-mo commented:
Thanks, everyone, for helping review and get this PR landed!
